pkg/reconciler/instances/istio/action_test.go

```diff
@@ -424,6 +424,142 @@ func newFakeKubeClient() *k8smocks.Client {
+		ClientVersion:    "1.0",
+		PilotVersion:     "1.0",
+		DataPlaneVersion: "1.0",
+	}
+	t.Run("should perform istio uninstall action when istio is available", func(t *testing.T) {
+		// given
+		factory := workspacemocks.Factory{}
+		provider := chartmocks.Provider{}
+		kubeClient := newFakeKubeClient()
+		actionContext := newFakeServiceContext(&factory, &provider, kubeClient)
+		performer := actionsmocks.IstioPerformer{}
+		performer.On("Version", mock.Anything, mock.AnythingOfType("string"), mock.AnythingOfType(
+			"string"), mock.AnythingOfType("string"), mock.AnythingOfType("*zap.SugaredLogger")).
+			Return(istioAvailable, nil)
+		performer.On("Uninstall", mock.Anything, mock.AnythingOfType("*zap.SugaredLogger")).Return(nil)
+		action := UninstallAction{performer: &performer}
+
+		// when
+		err := action.Run(actionContext)
+		// then
+		require.NoError(t, err)
+		performer.AssertCalled(t, "Version", mock.Anything, mock.AnythingOfType("string"), mock.
+			AnythingOfType("string"), mock.AnythingOfType("string"), mock.AnythingOfType("*zap.SugaredLogger"))
+		performer.AssertCalled(t, "Uninstall", mock.Anything, mock.AnythingOfType("*zap.SugaredLogger"))
+	})
+	t.Run("should not perform istio uninstall action when istio was not detected on the cluster", func(t *testing.T) {
+		// given
+		factory := workspacemocks.Factory{}
+		provider := chartmocks.Provider{}
+		kubeClient := newFakeKubeClient()
+		actionContext := newFakeServiceContext(&factory, &provider, kubeClient)
+		performer := actionsmocks.IstioPerformer{}
+		performer.On("Version", mock.Anything, mock.AnythingOfType("string"), mock.AnythingOfType(
+			"string"), mock.AnythingOfType("string"), mock.AnythingOfType("*zap.SugaredLogger")).
+			Return(noIstioOnTheCluster, nil)
+		action := UninstallAction{performer: &performer}
+
+		// when
+		err := action.Run(actionContext)
+		// then
+		require.NoError(t, err)
+		performer.AssertCalled(t, "Version", mock.Anything, mock.AnythingOfType("string"), mock.AnythingOfType("string"), mock.AnythingOfType("string"), mock.AnythingOfType("*zap.SugaredLogger"))
+		performer.AssertNotCalled(t, "Uninstall", mock.AnythingOfType("string"), mock.AnythingOfType("string"))
+	})
+	t.Run("should not perform istio uninstall action when there is an error detecting istio version", func(t *testing.T) {
+		// given
+		factory := workspacemocks.Factory{}
+		provider := chartmocks.Provider{}
+		kubeClient := newFakeKubeClient()
+		actionContext := newFakeServiceContext(&factory, &provider, kubeClient)
+		performer := actionsmocks.IstioPerformer{}
+		performer.On("Version", mock.Anything, mock.AnythingOfType("string"), mock.AnythingOfType(
+			"string"), mock.AnythingOfType("string"), mock.AnythingOfType("*zap.SugaredLogger")).
+			Return(noIstioOnTheCluster, errors.New("error in detecting istio version"))
+		action := UninstallAction{performer: &performer}
+
+		// when
+		err := action.Run(actionContext)
+		// then
+		require.Error(t, err)
```

Comment:
What about checking what was the error message returned? The fact that it exists may not be enough because some other struct could propagate it.

Reply:
Added